File: laser-prototype/js/geometry.js
var Geometry = (function() {
  return {
    createRectangle: function() {
      var halfWidth = 0.5;
      var halfHeight = 0.5;

      var vertices = [
        -halfWidth, -halfHeight,
         halfWidth, -halfHeight,
         halfWidth,  halfHeight,
        -halfWidth,  halfHeight
      ];

      var indices = [ 0, 1, 2, 3, 0 ];

      return {
        vertices: vertices,
        indices: indices
      };
    },

    createRing: function( options ) {
      options = options || {};

      var outerRadius   = options.outerRadius || 1.0,
          innerRadius   = options.innerRadius,
          startAngle    = options.startAngle,
          endAngle      = options.endAngle,
          subdivisions  = options.subdivisions || 32;
          anticlockwise = options.anticlockwise;

      // Allow false/zero values to be passed in.
      if ( typeof innerRadius === 'undefined' ) { innerRadius = 0.5; }
      if ( typeof startAngle === 'undefined' ) { startAngle = 0; }
      if ( typeof endAngle === 'undefined' ) { endAngle = 2 * Math.PI; }
      if ( typeof anticlockwise === 'undefined' ) { anticlockwise = true; }

      var vertices = [];
      var indices = [];

      var sweepAngle = endAngle - startAngle;

      // Check for circle.
      if ( Math.abs( sweepAngle ) >= 2 * Math.PI ) {
        endAngle = startAngle + 2 * Math.PI;
      } else {
        // Anticlockwise arcs have positive sweepAngles.
        // Clockwise arcs have negative sweepAngles.
        // If the above is false, move the sweepAngle to the correct sign.
        if ( anticlockwise && sweepAngle < 0 ) {
          sweepAngle += 2 * Math.PI;
        } else if ( !anticlockwise && sweepAngle > 0 ) {
          sweepAngle -= 2 * Math.PI;
        }
      }

      var subdivAngle = sweepAngle / subdivisions;

      // Outer radius.
      var i;
      for ( i = 0; i < subdivisions + 1; i++ ) {
        vertices.push( outerRadius * Math.cos( startAngle + i * subdivAngle ) );
        vertices.push( outerRadius * Math.sin( startAngle + i * subdivAngle ) );

        indices.push(i);
      }

      // Inner radius.
      for ( i = 0; i < subdivisions + 1; i++ ) {
        vertices.push( innerRadius * Math.cos( endAngle - i * subdivAngle ) );
        vertices.push( innerRadius * Math.sin( endAngle - i * subdivAngle ) );

        // The plus one takes into account the loop variable in outerRadius loop.
        indices.push( i + subdivisions + 1 );
      }

      // Edge connecting inner and outer radius.
      indices.push(0);

      return {
        vertices: vertices,
        indices: indices
      };
    }
  };
}) ();

File: laser/src/com/razh/laser/Geometry.java
package com.razh.laser;

import com.badlogic.gdx.graphics.Mesh;
import com.badlogic.gdx.graphics.VertexAttribute;
import com.badlogic.gdx.graphics.VertexAttributes.Usage;
import com.badlogic.gdx.graphics.glutils.ShaderProgram;

public class Geometry {
	public static Mesh createRing(float outerRadius, float innerRadius,
	                              float startAngle, float endAngle,
	                              int subdivisions, boolean anticlockwise) {

		// Two vertices on outer and inner radii with two components each.
		int vertexCount = (subdivisions + 1) * 4;
		int indexCount = (subdivisions + 1) * 2;

		Mesh mesh = new Mesh(Mesh.VertexDataType.VertexBufferObject,
		                     true, vertexCount, indexCount,
		                     new VertexAttribute(Usage.Position, 2,
		                                         ShaderProgram.POSITION_ATTRIBUTE));

		float[] vertices = new float[vertexCount];
		short[] indices = new short[indexCount];

		float sweepAngle = endAngle - startAngle;

		// Check if circle.
		if (Math.abs(sweepAngle) >= 2 * Math.PI) {
			endAngle = (float) (startAngle + 2 * Math.PI);
		} else {
			// Anticlockwise arcs have positive sweepAngles.
			// Clockwise arcs have negative sweepAngles.
			// If the the above is false, flip the sweepAngle to the correct sign.
			if (anticlockwise && sweepAngle < 0) {
				sweepAngle += 2 * Math.PI;
			} else if (!anticlockwise && sweepAngle > 0) {
				sweepAngle -= 2 * Math.PI;
			}
		}

		float subdivAngle = sweepAngle / subdivisions;

		int vtxIndex = 0;
		int idxIndex = 0;

		float cos, sin;
		for (int i = 0; i < subdivisions + 1; i++) {
			cos = (float) Math.cos(startAngle + i * subdivAngle);
			sin = (float) Math.sin(startAngle + i * subdivAngle);
			// Inner radius.
			vertices[vtxIndex++] = innerRadius * cos;
			vertices[vtxIndex++] = innerRadius * sin;
			// Outer radius.
			vertices[vtxIndex++] = outerRadius * cos;
			vertices[vtxIndex++] = outerRadius * sin;

			indices[idxIndex++] = (short) (2 * i);
			indices[idxIndex++] = (short) (2 * i + 1);
		}

		mesh.setVertices(vertices);
		mesh.setIndices(indices);

		return mesh;
	}

	public static Mesh createCircle(int subdivisions) {
		return createCircle(1.0f, subdivisions);
	}

	public static Mesh createCircle(float radius, int subdivisions) {
		// Subdivisions and center vertex with two components each.
		int vertexCount = (subdivisions + 1) * 2;
		// Center vertex and one connecting edge between start and end vertices.
		int indexCount = subdivisions + 2;

		Mesh mesh = new Mesh(Mesh.VertexDataType.VertexBufferObject,
		                     true, vertexCount, indexCount,
		                     new VertexAttribute(Usage.Position, 2,
		                                         ShaderProgram.POSITION_ATTRIBUTE));

		float subdivAngle = (float) (2 * Math.PI / subdivisions);

		float[] vertices = new float[vertexCount];
		short[] indices = new short[indexCount];

		int vtxIndex = 0;
		int idxIndex = 0;

		vertices[vtxIndex++] = 0.0f;
		vertices[vtxIndex++] = 0.0f;

		// Center.
		indices[idxIndex++] = 0;

		for (int i = 0; i < subdivisions; i++) {
			vertices[vtxIndex++] = (float) (radius * Math.cos(i * subdivAngle));
			vertices[vtxIndex++] = (float) (radius * Math.sin(i * subdivAngle));

			indices[idxIndex++] = (short) (i + 1);
		}

		// Close triangle fan.
		indices[idxIndex++] = 1;

		mesh.setVertices(vertices);
		mesh.setIndices(indices);

		return mesh;
	}

	public static GeometryData createRingHull(float outerRadius, float innerRadius,
	                                          float startAngle, float endAngle,
                                              int subdivisions, boolean anticlockwise) {
		float[] vertices = new float[(subdivisions + 1) * 4];
		// Two radii, plus connecting edge.
		short[] indices = new short[(subdivisions + 1) * 2 + 1];

		float sweepAngle = endAngle - startAngle;

		// Circle check.
		if (Math.abs(sweepAngle) >= 2 * Math.PI) {
			endAngle = (float) (startAngle + 2 * Math.PI);
		} else {
			if (anticlockwise && sweepAngle < 0) {
				sweepAngle += 2 * Math.PI;
			} else if (!anticlockwise && sweepAngle > 0) {
				sweepAngle -= 2 * Math.PI;
			}
		}

		float subdivAngle = sweepAngle / subdivisions;

		int vtxIndex = 0;
		int idxIndex = 0;

		// Outer radius.
		for (int i = 0; i < subdivisions + 1; i++) {
			vertices[vtxIndex++] = (float) (outerRadius * Math.cos(startAngle + i * subdivAngle));
			vertices[vtxIndex++] = (float) (outerRadius * Math.sin(startAngle + i * subdivAngle));

			indices[idxIndex++] = (short) i;
		}

		// Inner radius.
		for (int i = 0; i < subdivisions + 1; i++) {
			vertices[vtxIndex++]= (float) (innerRadius * Math.cos(endAngle - i * subdivAngle));
			vertices[vtxIndex++]= (float) (innerRadius * Math.cos(endAngle - i * subdivAngle));

			indices[idxIndex++] = (short) (i + subdivisions + 1);
		}

		// Edge connecting inner and outer radius.
		indices[idxIndex++] = 0;

		GeometryData data = new GeometryData();
		data.vertices = vertices;
		data.indices = indices;

		return data;
	}

	public static class GeometryData {
		public float[] vertices;
		public short[] indices;
	}
}
